src/navigation/failure_identification/watchdog.py

```diff
@@ -0,0 +1,69 @@
+import numpy as np
+import rospy
+from pandas import DataFrame
+
+# 10 or more data objects flagged as stuck, then you are actually stuck
+DF_THRESHOLD = 100
+STUCK_THRESHOLD = 50
+ANGULAR_THRESHOLD = 0.001
+LINEAR_THRESHOLD = 0.55
+
+
+class WatchDog:
+    def __init__(self, collector_in):
+        self.collector = collector_in
+
+    # def is_stuck(self, dataframe: DataFrame):
+    #     return False
+    def update_pointers(self):
+        self.collector.right_pointer = self.collector.row_counter
+        self.collector.left_pointer = self.collector.right_pointer - DF_THRESHOLD
+        rospy.logerr(f"RIGHT: {self.collector.right_pointer} LEFT: {self.collector.left_pointer}")
+
+    def check_angular_stuck(self, dataframe: DataFrame):
+        counter = 0
+        self.collector.left_pointer = self.collector.right_pointer - DF_THRESHOLD
+        while self.collector.left_pointer < self.collector.right_pointer:
+            actual = (dataframe.loc[self.collector._df["row"] == self.collector.left_pointer]["angular_velocity"]).iloc[
+                0
+            ]
+            commanded = (dataframe.loc[self.collector._df["row"] == self.collector.left_pointer]["cmd_vel_twist"]).iloc[
+                0
+            ]
+            # rospy.logerr(f"ANGULAR ACTUAL: {actual} COMMANDED: {commanded}")
+            if actual < ANGULAR_THRESHOLD and commanded > 0:
+                counter += 1
+            self.collector.left_pointer += 1
+        # rospy.logerr(f"ANGULAR STUCK COUNTER: {counter}")
+        return counter >= STUCK_THRESHOLD
+
+    def check_linear_stuck(self, dataframe: DataFrame):
+        counter = 0
+        self.collector.left_pointer = self.collector.right_pointer - DF_THRESHOLD
+        while self.collector.left_pointer < self.collector.right_pointer:
```

Comment:
I think it could look something like:

```
left = self.collector.left_pointer
right = self.collector.right_pointer
cmd = dataframe["cmd_vel_x"][left : right]
actual = dataframe["linear_velocity"][left : right]
num_failures = np.logical_and(actual < 0.5 , cmd > 0).sum()
return num_failures >= STUCK_THRESHOLD
```